### ISSUE TITLE
About the 3D visualization

### ISSUE THREAD

**Issue by liusida** [yuanming-hu/difftaichi#1]
In README, I saw the 3D visualization of `diffmpm3d.py` is super cool
https://raw.githubusercontent.com/yuanming-hu/public_files/master/learning/difftaichi/diffmpm3d.gif

but when I try to reproduce the result, the code only produced `.bin` files. I am wondering if that visualization was produced by other software?

Thank you.

**Comment by yuanming-hu** [comment]
That's rendered using https://github.com/yuanming-hu/difftaichi/blob/master/examples/diffmpm_renderer.py 

However, this script is currently broken due to a recent compiler backend update. I'll fix this within the next week and please have fun with other examples. Sorry about that.

**Comment by liusida** [comment]
> That's rendered using https://github.com/yuanming-hu/difftaichi/blob/master/examples/diffmpm_renderer.py
> 
> However, this script is currently broken due to a recent compiler backend update. I'll fix this within the next week and please have fun with other examples. Sorry about that.

Thanks. This project is amazing, and I can feel that you are doing most of the work. Hope there will be more people to join and help, especially on document and source code management.  :laughing:

**Comment by liusida** [comment]
Hi! :hand:  just a reminder, the file "diffmpm_renderer.py" still produces this error:

```
No module named 'renderer_utils'
```

**Comment by masahi** [comment]
Hi I also want to see some pretty pictures. Is this something that a complete beginner to taichi can take on? I see the visualization itself is written in taichi :)

**Comment by yuanming-hu** [comment]
Hi guys, sorry about the delay. I'm recently refactoring Taichi for more compatibility and performance. I'll fix the renderer within 3 days. (Ping me again if I don't, thanks.)

**Comment by masahi** [comment]
ping? No hurry for me, of course.

**Comment by yuanming-hu** [comment]
oh sorry i've got too many things to do. Will try tonight/early next week.

**Comment by yuanming-hu** [comment]
For those who are eager to render the particles: the working demo is https://github.com/taichi-dev/taichi/blob/cf77f936e3c084003dc80067fbdcba9762cfa5e3/examples/renderer.py

**Comment by liusida** [comment]
> For those who are eager to render the particles: the working demo is https://github.com/taichi-dev/taichi/blob/cf77f936e3c084003dc80067fbdcba9762cfa5e3/examples/renderer.py

Thank you~

**Comment by masahi** [comment]
> For those who are eager to render the particles: the working demo is https://github.com/taichi-dev/taichi/blob/cf77f936e3c084003dc80067fbdcba9762cfa5e3/examples/renderer.py

hmm I tried examples/renderer.py with the latest taichi but it seems it is not an animation? Is this the expected result?

![Screenshot from 2020-02-16 07-43-57](https://user-images.githubusercontent.com/1776403/74596282-6815e700-5090-11ea-969d-4543dc10bd2b.png)

**Comment by yuanming-hu** [comment]
Yeah this is just a bunch of random particles within a cube. Particle loading is not yet implemented. (Sorry I've got too many things to do recently...the particle output of diffmpm3d.py is recently updated for performance (https://github.com/yuanming-hu/difftaichi/commit/1e9c804cf721a9c02301c7ff29d92cc8df0e14f6), yet I haven't got a chance to update the renderer to load it)

**Comment by masahi** [comment]
> Sorry I've got too many things to do recently

yeah I can imagine that, no problem :)

**Comment by zlyfunction** [comment]
Hi yuanming, is the visualization part working now? Thx

**Comment by KineticPayload** [comment]
Hello! I'm also trying to visualize the diffmpm3d simulation example. Has there been any progress on this issue with the renderer? Is there currently a way to visualize the output .bin files?

**Comment by KraftOreo** [comment]
Hi Yuanming, I just tried out the diffmpm3d.py demo and others. The 2d diffmpm is truly remarkable, but still I cannot visualize the results of the 3d version. It's been another three years. Is there any chance that I can get a bin file renderer or that I can find some documentation about loading and rendering the bin files by myself? Thanks in advance for all your efforts! Looking forward to your reply!

**Comment by abhayKashyap03** [comment]
Hey @yuanming-hu , just a follow-up/reminder ping asking for a visualizer/renderer for the 3D demo. Would be cool to run it locally instead of just seeing the gif on the repo/youtube video.
